internal/client.go

```diff
@@ -1003,6 +1024,26 @@ func DialCloudOperationsClient(ctx context.Context, options CloudOperationsClien
 	}, nil
 }
 
+// PrepareUpdateWorkflowOperation returns an UpdateWorkflowOperation that can be used to perform Update-with-Start,
+// or an error in case the update operation is invalid.
+func PrepareUpdateWorkflowOperation(options UpdateWorkflowOptions) (*UpdateWorkflowOperation, error) {
+	input, err := createUpdateWorkflowInput(options)
+	if err != nil {
+		return nil, err
+	}
+	return &UpdateWorkflowOperation{input: input}, nil
+}
+
+// Get blocks on the outcome of the update *after* it was executed via Client.ExecuteWorkflow.
+func (op *UpdateWorkflowOperation) Get(ctx context.Context, valuePtr interface{}) error {
+	if op.WorkflowUpdateHandle == nil {
+		return fmt.Errorf("update operation has not been executed")
+	}
+	return op.WorkflowUpdateHandle.Get(ctx, valuePtr)
+}
```

Comment:
I got here after some trial and error. Options I've considered:

**(1) `Get(run WorkflowRun) (WorkflowUpdateHandle, error)`**

The safest but most verbose option. The `run` parameter can be used to ensure the function is invoked _after_ `ExecuteWorkflow`. It could also verify that the update handle's WorkflowId and RunId match the run's (which doesn't 100% guarantee that they were used together, but it's probably good enough). But all that requires an extra error check that is _only_ there to prevent a mistake by the user, not to surface an actual client/server error. So users should be able to catch this in testing already.

```go
updHandle, err := upd.Get(run)
if err != nil {
    return nil, err
}

var result any
if err := updHandle.GetResult(context.Background() &result); err != nil {
    return nil, err 
}
``` 

**(1.1) `GetResult(ctx context.Context, run WorkflowRun, valuePtr interface{}) error`**

Combined with (1), adding an additional shortcut to get the update outcome directly compresses two calls into one. This is based on the assumption that in most cases, users care about the result more than anything else from the `WorkflowUpdateHandle` (but they could still obtain it, if needed).

```go
var result any
if err := upd.GetResult(context.Background(), run, &result); err != nil {
    return nil, err 
}
``` 

**(2) `Handle(run WorkflowRun) WorkflowUpdateHandle`**

Removes the error check from (1). If the user passes in `nil`, it would panic (akin to a NPE since this is a user error). If the WorkflowId/RunId don't match, it could panic/ignore that (far from ideal). Overall my least favorite option.

```go
var result any
if err := upd.Handle(run).Get(context.Background(), &result); err != nil {
    return nil, err 
}
``` 

**(3) `Get(ctx context.Context, valuePtr interface{}) error`** (as written here)

When the operation embeds `WorkflowUpdateHandle`, the extra step to obtain the handle is removed entirely. While this is the most convenient option, it also means that there is a risk the user invokes the handle's functions _before_ executing the operation (however, if `Get` is invoked too early, it returns the appropriate error - the same cannot be said for WorkflowID and RunID, which would be empty).

```go
var result any
if err := upd.Get(context.Background(), &result); err != nil {
    return nil, err 
}
``` 

---


(1) is what we discussed in our design sessions. After implementing it, it felt unnecessarily complex/verbose to me compared to (4). But I do see the risk that introduces, too. Tell me what you think.

Reply:
Okay, for some reason I was approached this with the single-threaded "inside a workflow" mindset. The "wait for handle" use case makes perfect sense. I made it concurrency-safe now.